### PR TITLE
WIP Implement nocopy support for bincode

### PR DESCRIPTION
**Pull request by TyOverby**
TODO:

* [x] Find origina of test failure and fix it
* [x] Make deserialize_str and deserialize_bytes behave correctly in the face of a SizeLimit
* [x] Make BincodeRead public but prevent implementation
* [x] Stop using pub(crate)

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -14,8 +14,8 @@ description = "A binary serialization / deserialization strategy that uses Serde
 [dependencies]
 byteorder = "1.0.0"
 num-traits = "0.1.32"
-serde = { git = "https://github.com/serde-rs/serde", branch = "master" }
+serde = "1.0.*"
```
Comment (dtolnay) [inline Cargo.toml]: This disallows upgrading to 1.1.0 which is not a breaking change. I think this should say "1.0".

File: src/de/mod.rs [hunk 2]
```diff
@@ -8,6 +8,9 @@ use serde_crate::de::IntoDeserializer;
 use serde_crate::de::Error as DeError;
 use ::SizeLimit;
 use super::{Result, Error, ErrorKind};
+use self::read::BincodeRead;
+
+pub(crate) mod read;
```
Comment (dtolnay) [inline src/de/mod.rs]: `pub(crate)` doesn't work on stable yet so I would strongly discourage you from releasing with it.

**Comment by dtolnay** [comment]
Other than that, and the checklist, I think that does it. Not so much code after all!

File: src/de/mod.rs (hunk 2 above)
Comment (TyOverby) [inline src/de/mod.rs]: Ah yep, this was mainly added just to just so I wouldn't have to think about it when I was prototyping the new module structure.  I'll add this to my list of TODOs at the PR description.

**Comment by TyOverby** [comment]
@dtolnay Yeah, I was pretty surprised at how easy it was!  (I think you helped out a lot with your initial 1.0 change though)

File: Cargo.toml [hunk 3]
```diff
@@ -14,8 +14,8 @@ description = "A binary serialization / deserialization strategy that uses Serde
 [dependencies]
 byteorder = "1.0.0"
 num-traits = "0.1.32"
-serde = { git = "https://github.com/serde-rs/serde", branch = "master" }
+serde = "1.*.*"
```
Comment (cedenday) [inline Cargo.toml]: I don't like this, you should just write "^1", or more simply "1".

Comment (TyOverby) [inline Cargo.toml]: This is more explicit and easier to read.

**Comment by TyOverby** [comment]
@dtolnay: I think this one is ready to merge in!

File: src/lib.rs [hunk 4]
```diff
@@ -40,30 +40,32 @@ extern crate byteorder;
 extern crate num_traits;
 extern crate serde as serde_crate;
 
-mod serde;
+mod ser;
+mod de;
+pub mod internal;
```
Comment (dtolnay) [inline src/lib.rs]: This is not supposed to be public right?

Comment (TyOverby) [inline src/lib.rs]: I renamed `endian_choice` to `internal` and kept it public.  I'm debating if I should switch back to the old name though.

**Comment by heartsucker** [comment]
Can you guys push this to crates.io? I'd like to be able to use this in my libs and not depend on git commits.

**Comment by TyOverby** [comment]
@heartsucker pretty sure this is already published as 0.8.0

**Pull request merged**